src/lib/core/media-marshaller/media-marshaller.ts

```diff
@@ -0,0 +1,218 @@
+   * @param observables
+   */
+  init(element: HTMLElement,
+       key: string,
+       builder?: Function,
+       observables: Observable<any>[] = []): void {
+    if (builder) {
+      let builders = this.builderMap.get(element);
+      if (!builders) {
+        builders = new Map();
+        this.builderMap.set(element, builders);
+      }
+      builders.set(key, builder);
+    }
+    if (observables) {
+      let watchers = this.watcherMap.get(element);
+      if (!watchers) {
+        watchers = new Map();
+        this.watcherMap.set(element, watchers);
+      }
+      const subscription = watchers.get(key);
+      if (!subscription) {
+        const newSubscription = merge(...observables).subscribe(() => {
+          const currentValue = this.getValue(element, key);
+          this.updateElement(element, key, currentValue);
+        });
+        watchers.set(key, newSubscription);
+      }
+    }
+  }
+
+  /**
+   * get the value for an element and key and optionally a given breakpoint
+   * @param element
+   * @param key
+   * @param bp
+   */
+  getValue(element: HTMLElement, key: string, bp?: string): string {
+    const bpMap = this.elementMap.get(element);
+    if (bpMap) {
+      const values = bp !== undefined ? bpMap.get(bp) : this.getFallback(bpMap);
+      if (values) {
+        const value = values.get(key);
+        return value !== undefined ? value : '';
+      }
+    }
+    return '';
+  }
+
+  /**
+   * whether the element has values for a given key
+   * @param element
+   * @param key
+   */
+  hasValue(element: HTMLElement, key: string): boolean {
+    const bpMap = this.elementMap.get(element);
+    if (bpMap) {
+      const values = this.getFallback(bpMap);
+      if (values) {
+        return values.get(key) !== undefined || false;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Set the value for an input on a directive
+   * @param element the element in question
+   * @param key the type of the directive (e.g. flex, layout-gap, etc)
+   * @param bp the breakpoint suffix (empty string = default)
+   * @param val the value for the breakpoint
+   */
+  setValue(element: HTMLElement, key: string, val: any, bp: string): void {
+    let bpMap: BreakpointMap | undefined = this.elementMap.get(element);
+    if (!bpMap) {
+      bpMap = new Map().set(bp, new Map().set(key, val));
+      this.elementMap.set(element, bpMap);
+    } else {
+      const values = (bpMap.get(bp) || new Map()).set(key, val);
+      bpMap.set(bp, values);
+      this.elementMap.set(element, bpMap);
+    }
+    this.updateElement(element, key, this.getValue(element, key));
+  }
+
+  trackValue(element: HTMLElement, key: string): Observable<string> {
+    let elementMap = this.subjectMap.get(element);
+    if (!elementMap) {
+      elementMap = new Map();
+      this.subjectMap.set(element, elementMap);
+    }
+    let subject = elementMap.get(key);
+    if (!subject) {
+      subject = new Subject();
```

Comment:
We could use a single Subject and filter by key/element. Then the single subject becomes the central stream for all events.